File: tests/merge/last-wins.test.ts
import { describe, expect, it } from "vitest";

import { lastWinsStrategy } from "../../src/merge/last-wins.js";

describe("lastWinsStrategy (R10)", () => {
  it("returns the last contributor's bytes verbatim", () => {
    const r = lastWinsStrategy("commands/x.sh", [
      { id: "a", bytes: Buffer.from("from-a\n") },
      { id: "b", bytes: Buffer.from("from-b\n") },
    ]);
    expect(r.bytes.toString("utf8")).toBe("from-b\n");
    expect(r.contributors).toEqual(["b"]);
  });

  it("single-contributor case is the trivial pass-through", () => {
    const buf = Buffer.from("solo");
    const r = lastWinsStrategy("agents/foo.json", [{ id: "only", bytes: buf }]);
    expect(Buffer.compare(r.bytes, buf)).toBe(0);
    expect(r.contributors).toEqual(["only"]);
  });

  it("returns a fresh Buffer (no aliasing of the input)", () => {
    // Future callers (E3 dry-run, E5 validate) must be able to mutate the
    // output without corrupting input bytes the orchestrator may still hold.
    const buf = Buffer.from("payload");
    const r = lastWinsStrategy("agents/foo.json", [{ id: "only", bytes: buf }]);
    expect(r.bytes).not.toBe(buf);
    r.bytes.fill(0);
    expect(buf.toString("utf8")).toBe("payload");
  });

  it("ignores all but the last contributor in provenance", () => {
    const r = lastWinsStrategy("plugin.json", [
      { id: "x1", bytes: Buffer.from("1") },
      { id: "x2", bytes: Buffer.from("2") },
      { id: "x3", bytes: Buffer.from("3") },
    ]);
    expect(r.bytes.toString()).toBe("3");
    expect(r.contributors).toEqual(["x3"]);
  });

  it("throws if invoked with no inputs", () => {
    expect(() => lastWinsStrategy("foo.txt", [])).toThrow();
  });
});

File: internal/merge/last_wins_test.go
package merge

import (
	"bytes"
	"testing"
)

// R10: last-wins picks the highest contributor index and lists only that
// contributor in provenance.
func TestLastWinsStrategy_LastBytesWin(t *testing.T) {
	r, err := LastWinsStrategy("commands/x.sh", []ContributorBytes{
		{ID: "a", Bytes: []byte("from-a\n")},
		{ID: "b", Bytes: []byte("from-b\n")},
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if string(r.Bytes) != "from-b\n" {
		t.Fatalf("bytes: want %q, got %q", "from-b\n", string(r.Bytes))
	}
	if len(r.Contributors) != 1 || r.Contributors[0] != "b" {
		t.Fatalf("contributors: want [b], got %v", r.Contributors)
	}
}

func TestLastWinsStrategy_SinglePassthrough(t *testing.T) {
	in := []byte("solo")
	r, err := LastWinsStrategy("agents/foo.json", []ContributorBytes{{ID: "only", Bytes: in}})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !bytes.Equal(r.Bytes, in) {
		t.Fatalf("bytes diverged: want %q, got %q", in, r.Bytes)
	}
	if len(r.Contributors) != 1 || r.Contributors[0] != "only" {
		t.Fatalf("contributors: want [only], got %v", r.Contributors)
	}
}

// Future callers (D5, validate) must be able to mutate the output without
// corrupting input bytes still held by the orchestrator.
func TestLastWinsStrategy_FreshBufferNotAliased(t *testing.T) {
	in := []byte("payload")
	r, err := LastWinsStrategy("agents/foo.json", []ContributorBytes{{ID: "only", Bytes: in}})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	for i := range r.Bytes {
		r.Bytes[i] = 0
	}
	if string(in) != "payload" {
		t.Fatalf("input was mutated by output zeroing: %q", string(in))
	}
}

func TestLastWinsStrategy_OnlyLastInProvenance(t *testing.T) {
	r, err := LastWinsStrategy("plugin.json", []ContributorBytes{
		{ID: "x1", Bytes: []byte("1")},
		{ID: "x2", Bytes: []byte("2")},
		{ID: "x3", Bytes: []byte("3")},
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if string(r.Bytes) != "3" {
		t.Fatalf("bytes: want %q, got %q", "3", string(r.Bytes))
	}
	if len(r.Contributors) != 1 || r.Contributors[0] != "x3" {
		t.Fatalf("contributors: want [x3], got %v", r.Contributors)
	}
}

func TestLastWinsStrategy_NoInputsErrors(t *testing.T) {
	if _, err := LastWinsStrategy("foo.txt", nil); err == nil {
		t.Fatal("want error for empty inputs, got nil")
	}
}

// A zero-byte winner must still produce a valid result with empty bytes
// and the winner in provenance — distinct from "no inputs" (which errors).
func TestLastWinsStrategy_ZeroByteWinner(t *testing.T) {
	r, err := LastWinsStrategy("plugin.json", []ContributorBytes{
		{ID: "a", Bytes: []byte("from-a\n")},
		{ID: "empty", Bytes: []byte{}},
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(r.Bytes) != 0 {
		t.Fatalf("bytes: want empty, got %q", string(r.Bytes))
	}
	if len(r.Contributors) != 1 || r.Contributors[0] != "empty" {
		t.Fatalf("contributors: want [empty], got %v", r.Contributors)
	}
}
